File: pfm-haskell-twain/explore/Similarity.hs
{-# LANGUAGE OverloadedStrings #-}

module Similarity where

import Data.Char (isAlphaNum, toLower)
import Data.Function (on)
import Data.List (sortBy)
import Data.Set qualified as S
import System.IO

-- | Normalize a string by removing non-alphanum chars and lowercasing
normalize :: String -> String
normalize = map toLower . filter (\c -> isAlphaNum c || c == ' ')

-- | Tokenize a string into a set of lowercase words
tokenize :: String -> S.Set String
tokenize = S.fromList . words . normalize

-- | Jaccard similarity between two token sets
jaccard :: S.Set String -> S.Set String -> Double
jaccard a b
    | S.null union = 0
    | otherwise = fromIntegral (S.size inter) / fromIntegral (S.size union)
  where
    inter = S.intersection a b
    union = S.union a b

-- | Naive similarity: number of common words
naiveOverlap :: S.Set String -> S.Set String -> Int
naiveOverlap a b = S.size $ S.intersection a b

-- | Levenshtein distance (edit distance)
levenshtein :: String -> String -> Int
levenshtein s t = last (foldl transform [0 .. length s] t)
  where
    transform xs@(x : xs') c = scanl compute (x + 1) (zip3 s xs xs')
      where
        compute z (c', x', y') = minimum [y' + 1, z + 1, x' + fromEnum (c' /= c)]

-- | Sample transactions (raw descriptions)
transactions :: [String]
transactions =
    [ "ORANGE SA-ORANGE"
    , "ORANGE MOBILE BILL"
    , "NETFLIX 12/06"
    , "SPAR PARIS 10/05"
    , "SPOTIFY FAMILY PLAN"
    , "AMAZON FRANCE SAS"
    ]

main :: IO ()
main = do
    putStrLn "Enter a transaction description to compare:"
    hFlush stdout
    input <- getLine

    let inputTok = tokenize input

    putStrLn "\n--- Similarity Rankings ---\n"

    let scored =
            map
                ( \s ->
                    ( s
                    , jaccard inputTok (tokenize s)
                    , naiveOverlap inputTok (tokenize s)
                    , levenshtein input s
                    )
                )
                transactions

    putStrLn "By Jaccard similarity:"
    mapM_ print $ take 3 $ reverse $ sortOn (\(_, score, _, _) -> score) scored

    putStrLn "\nBy naive word overlap:"
    mapM_ print $ take 3 $ reverse $ sortOn (\(_, _, overlap, _) -> overlap) scored

    putStrLn "\nBy Levenshtein distance (lower is better):"
    mapM_ print $ take 3 $ sortOn (\(_, _, _, dist) -> dist) scored

-- Helper: sort list using a function
sortOn :: (Ord b) => (a -> b) -> [a] -> [a]
sortOn f = map snd . sortPairs . map (\x -> (f x, x))
  where
    sortPairs = sortBy (compare `on` fst)

File: pfm-haskell-twain/explore/similarity.rb
def normalize(str)
  str.downcase.gsub(/[^a-z0-9 ]/, '')
end

def tokenize(str)
  normalize(str).split.uniq.to_set
end

def jaccard(a, b)
  return 0.0 if (a | b).empty?
  (a & b).size.to_f / (a | b).size
end

def naive_overlap(a, b)
  (a & b).size
end

def levenshtein(a, b)
  a_len = a.length
  b_len = b.length
  d = Array.new(a_len + 1) { |i| [i] + [0] * b_len }

  (1..b_len).each { |j| d[0][j] = j }

  (1..a_len).each do |i|
    (1..b_len).each do |j|
      cost = a[i - 1] == b[j - 1] ? 0 : 1
      d[i][j] = [
        d[i - 1][j] + 1,
        d[i][j - 1] + 1,
        d[i - 1][j - 1] + cost
      ].min
    end
  end

  d[a_len][b_len]
end

require 'set'

TRANSACTIONS = [
  "ORANGE SA-ORANGE",
  "ORANGE MOBILE BILL",
  "NETFLIX 12/06",
  "SPAR PARIS 10/05",
  "SPOTIFY FAMILY PLAN",
  "AMAZON FRANCE SAS"
]

print "Enter a transaction description to compare:\n> "
input = gets.chomp

input_tokens = tokenize(input)

scored = TRANSACTIONS.map do |t|
  tokens = tokenize(t)
  [
    t,
    jaccard(input_tokens, tokens),
    naive_overlap(input_tokens, tokens),
    levenshtein(normalize(input), normalize(t))
  ]
end

puts "\nBy Jaccard similarity:"
scored.sort_by { |_, sim, _, _| -sim }.first(3).each { |row| p row }

puts "\nBy naive word overlap:"
scored.sort_by { |_, _, overlap, _| -overlap }.first(3).each { |row| p row }

puts "\nBy Levenshtein distance (lower is better):"
scored.sort_by { |_, _, _, dist| dist }.first(3).each { |row| p row }


require 'pry'; binding.pry
